core/server/common/src/main/java/alluxio/master/MasterRegistry.java

```diff
@@ -116,9 +95,56 @@ public MasterRegistry() {}
   /**
    * @return a collection of all the registered masters
    */
-  public synchronized Collection<Master> getMasters() {
-    List<Master> masters = new ArrayList<>(mRegistry.values());
-    Collections.sort(masters, COMPARATOR);
+  public Collection<Master> getMasters() {
+    List<Map.Entry<Class<?>, Master>> entries = new ArrayList<>(mRegistry.entrySet());
+    Collections.sort(entries, new DependencyComparator());
+    List<Master> masters = new ArrayList<>();
+    for (Map.Entry<Class<?>, Master> entry : entries) {
+      masters.add(entry.getValue());
+    }
     return masters;
   }
+
+  private Set<Class<?>> getTransitiveDeps(Class<?> clazz) {
```

Comment:
classes are the keys and the dependency graph is expressed in terms of the classes not in terms of the objects ...

are you suggesting that `getDependecies()`  returns a set of masters? if so, that would impose limitation on when it can be initialized / computed (it would need to happen after the dependencies are passed into the constructor)